### PR TITLE
Robust containerized externalbrowser rebased

### LINKED ISSUE #1251: SNOW-660632: Better support for EXTERNALBROWSER auth from local containers
## What is the current behaviour?

EXTERNALBROWSER authentication is the best option in the case of SSO with MFA. However, when performed inside a local container such as a VSCode docker devcontainer, there is a race condition between the auth callback to the localhost and starting the receiver app on the localhost and auto-forwarding the randomly assigned port. This can lead to an intermittent failure to connect.

One way to address this is to assign a fixed port number with the SF_AUTH_SOCKET_PORT environment variable and statically forward this port. This removes the delay in detecting and autoforwarding a randomly assigned port. However a new problem arises: the port remains in a TIME_WAIT state for a few minutes after it is used and subsequent connections fail with `OSError: [Errno 98] Address already in use` during this period.

## What is the desired behaviour?

EXTERNALBROWSER auth works every time from inside a local container.

An option is to set up a static port forwarding rule as above and set the `SO_REUSEADDR` state of the port, perhaps determined by the value of new environment variable such as `SF_AUTH_SOCKET_REUSEADDR`.

## How would this improve `snowflake-connector-python`?

EXTERNALBROWSER auth would work more seamlessly from local containers.

## References, Other Background

[socket(7) — Linux manual page](https://man7.org/linux/man-pages/man7/socket.7.html)
[SF_AUTH_SOCKET_PORT code](https://github.com/snowflakedb/snowflake-connector-python/blob/0b24148ec0c947764c50f093f6bc1fe1638fb551/src/snowflake/connector/auth_webbrowser.py#L103)

I'm up for doing a PR to address this.

### PR DESCRIPTION
**Pull request by raulbonet**
Please answer these questions before submitting your pull requests. Thanks!

Note: Copy&Pasting the original message from @podung , in his original PR [here](https://github.com/snowflakedb/snowflake-connector-python/pull/1686)

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #1251 

2. Fill out the following pre-review checklist:

   - [X] I am adding a new automated test(s) to verify correctness of my new code
   - [X] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [X] I am modifying authorization mechanisms
   - [ ] I am adding new credentials
   - [ ] I am modifying OCSP code
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

* Improved `externalbrowser` auth in containerized environments
    * Instruct browser to not fetch `/favicon `on success page (can cause failure to parse subsequent tokens)
    * Simple retry strategy on empty socket.recv (occurs frequently when containerized)
    * Add `SF_AUTH_SOCKET_REUSEPORT` flag (usage: `SF_AUTH_SOCKET_REUSEPORT=true`) to set the underlying socket's `SO_REUSEPORT` flag (described in the [socket man page](https://man7.org/linux/man-pages/man7/socket.7.html))
        * Use in combination with `SF_AUTH_SOCKET_PORT`
        * Consider using if running in a containerized environment and the localhost callback url is being followed before the container fully opens the port
    * Add `SF_AUTH_SOCKET_MSG_DONTWAIT` flag (usage: `SF_AUTH_SOCKET_MSG_DONTWAINT=true`) to make a non-blocking socket.recv call and retry on Error
        * Consider using this if running in a containerized environment and externalbrowser auth frequently hangs while waiting for callback
        * NOTE: this has not been tested extensively, but has been shown to improve the experience when using WSL

### CODE REVIEW

**Comment by raulbonet** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by raulbonet** [comment]
recheck

**Comment by raulbonet** [comment]
Hello @sfc-gh-mkeller @sfc-gh-aalam 
Could somebody approve the workflows to trigger the tests?
Thank you

File: test/unit/test_auth_webbrowser.py [hunk 1]
```diff
@@ -455,7 +455,7 @@ def test_auth_webbrowser_invalid_sso(monkeypatch):
 def test_auth_webbrowser_socket_recv_retries_up_to_15_times_on_empty_bytearray():
     """Authentication by WebBrowser retries on empty bytearray response from socket.recv"""
     ref_token = "MOCK_TOKEN"
-    rest = _init_rest(INVALID_SSO_URL, REF_PROOF_KEY, disable_console_login=True)
```
Comment (raulbonet) [inline test/unit/test_auth_webbrowser.py]: @podung , if I am understanding correctly this test, we do need a valid SSO URL here

File: test/unit/test_auth_webbrowser.py [hunk 2]
```diff
@@ -540,7 +540,7 @@ def test_auth_webbrowser_socket_recv_does_not_block_with_env_var(monkeypatch):
     """Authentication by WebBrowser socket.recv Does not block, but retries if BlockingIOError thrown."""
 
     ref_token = "MOCK_TOKEN"
-    rest = _init_rest(REF_SSO_URL, REF_PROOF_KEY)
+    rest = _init_rest(REF_SSO_URL, REF_PROOF_KEY, disable_console_login=True)
```
Comment (raulbonet) [inline test/unit/test_auth_webbrowser.py:543]: @podung  In a similar way to the rest of the tests, indeed it seems that now the proof_key is encoded when `disable_console_login=False`. For now I just set it to true to make it work, but we can use an if-condition like happens in other tests

File: test/unit/test_auth_webbrowser.py [hunk 3]
```diff
@@ -455,7 +455,7 @@ def test_auth_webbrowser_invalid_sso(monkeypatch):
 def test_auth_webbrowser_socket_recv_retries_up_to_15_times_on_empty_bytearray():
     """Authentication by WebBrowser retries on empty bytearray response from socket.recv"""
     ref_token = "MOCK_TOKEN"
-    rest = _init_rest(INVALID_SSO_URL, REF_PROOF_KEY, disable_console_login=True)
+    rest = _init_rest(REF_SSO_URL, REF_PROOF_KEY, disable_console_login=True)
```
Comment (raulbonet) [inline test/unit/test_auth_webbrowser.py:458]: @podung , if I am understanding correctly this test, we do need a valid SSO URL here

**Comment by raulbonet** [comment]
Hello @sfc-gh-aalam 
Not sure why the Build tasks were not triggered.
In any case, I think the PR is ready for review. I have not done any major changes respect to the original PR that @podung already worked on.
Can somebody have a look?

**Comment by raulbonet** [comment]
Thanks @sfc-gh-aalam . Some tests are just not applicable in Windows, should be fixed.

File: test/unit/test_auth_webbrowser.py [hunk 4]
```diff
@@ -393,3 +453,312 @@ def test_auth_webbrowser_invalid_sso(monkeypatch):
...
+        assert auth.assertion_content is None
+        assert sleep.call_count == 0
+
+
+def test_auth_webbrowser_socket_recv_does_not_block_with_env_var(monkeypatch):
+    """Authentication by WebBrowser socket.recv Does not block, but retries if BlockingIOError thrown."""
+
+    if not IS_WINDOWS:
+        # MSG_DONTWAIT is not supported on Windows, so this case is not applicable
+        assert True
```
[the remark this answers is not in the document]
Comment (raulbonet) [reply test/unit/test_auth_webbrowser.py]: Totally true, forgot about that one, fixed! @sfc-gh-aalam

**Comment by raulbonet** [comment]
Hello @sfc-gh-aalam 
Do you need anything else from me? I see some tests failing but I am not sure it is related to the development

File: test/unit/test_auth_webbrowser.py (hunk 2 above)
Comment (podung) [reply test/unit/test_auth_webbrowser.py:546]: Sorry for the delays here.  Yep that sounds exactly right to me

File: test/unit/test_auth_webbrowser.py (hunk 3 above)
Comment (podung) [reply test/unit/test_auth_webbrowser.py:461]: Yes - that looks like a result of a poor rebase conflict resolution on my part.  I agree 100%

**Review by sfc-gh-aalam — APPROVED** [review]

**Comment by sfc-gh-aalam** [comment]
Thank @raulbonet @podung for you hard work on this PR!

**Pull request merged**